GVFS/GVFS.Common/Tracing/JsonTracer.cs

```diff
@@ -339,40 +340,39 @@ private void WriteEvent(string eventName, EventLevel level, Keywords keywords, E
 
         private void MarkAndLogListenerRecovery(EventListener recoveredListener)
         {
+            // Check ContainsKey first (rather than always calling TryRemove) because ContainsKey
+            // is lock-free and recoveredListener should rarely be in failedListeners
             if (!this.failedListeners.ContainsKey(recoveredListener))
             {
-                // This listener has not failed since the last time it was called, so no need to do anything
+                // This listener has not failed since the last time it was called, so no need to log recovery
                 return;
             }
 
-            this.failedListeners.Remove(recoveredListener);
-
-            TraceEventMessage message = CreateListenerRecoveryMessage(recoveredListener);
-
-            // Only log that the listener has recovered to the other good listeners
-            foreach (EventListener listener in this.listeners.Except(this.failedListeners.Keys))
+            if (this.failedListeners.TryRemove(recoveredListener, out _))
             {
-                listener.TryRecordMessage(message, out _);
+                TraceEventMessage message = CreateListenerRecoveryMessage(recoveredListener);
+
+                // Only log that the listener has recovered to the other good listeners
+                foreach (EventListener listener in this.listeners.Except(this.failedListeners.Keys))
+                {
+                    listener.TryRecordMessage(message, out _);
+                }
             }
         }
 
         private void MarkAndLogListenerFailure(EventListener failedListener, string errorMessage)
         {
-            if (this.failedListeners.ContainsKey(failedListener))

```

Comment:
For anyone else reviewing the Contains key is removes here cause there is no perf win and it's unnecessary with TryAdd.